Add newsletter and dois (#89)
drmowinckels/drmowinckels.github.io: content/blog/2024/09-01_fair_blog/index.qmd

zenodo_api_token <- Sys.getenv("ZENODO_API_TOKEN")
```

The Zenodo api endpoint is where you connect to all the funcionality of Zenodo you can access programatically. 
This is the base URL for any other URL we will construct.

The `ZENODO_API_TOKEN` is a token you can grab from your Zenodo profile that allows you to use their API by authenticating you. 
I have this stored in my `.Renviron` file, which I **do not** commit to git.
Now, make **very sure** you do not commit a plain-text file with an API token in it to git, as it exposes your personal token to the world. 
It's like telling people your credit card number, anyone can use it, and you don't want that.
In stead, we add this token also in the [Secrets section](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions) of Github settings for the repository in question, and the script will be able to pick it up from there.

This first thing we will do, is upload the meta-data to Zenodo.
This initiates an archive, and we can add the pdf to it later.
Using {httr2} we can build a curl request to the Zenodo API so that our archive is initiated.

> There are (usually) three main parts of most quaries to API's,

Reviewer: ```suggestion
There are (usually) three main parts of most queries to API's,
```

1. the API URL to query
1. an authenticating part
1. something we sent to the API (the body)

> In {httr2} this process is done through a series of steps that we usually will pipe into eachother like below.

Reviewer: ```suggestion
In {httr2} this process is done through a series of steps that we usually will pipe into each other like below.
```

We need to provide our token as before, and provide the actual file we want to upload with `req_body_file`.
I did this for so many attempts and could not get the API to accept my request.

I turned to the rOpenSci folks for help, I can always trust someone in there has tried what I'm trying before.
The last piece of the puzzle was switching to the `PUT` curl method, rather than the default `POST` method.
I had been looking at the curl commands needed as stated on the Zenodo docs, and just completely missed the fact that it was a `PUT` method rather than a `POST` method. 

```r
# Upload the pdf file
upload_response <- request(deposition$links$bucket) |> 
  req_url_path_append(pdf_file) |> 
  req_auth_bearer_token(zenodo_api_token) |> 
  req_body_file(pdf_file) |> 
  req_method("PUT") |> 
  req_perform()
```

We are looking for another positive result of that, like a 201 again.
Once we have that, our last piece for httr2 code is to actually publish the archive. 
Everything we have done untill now can still be undone.

> Meaning we can delete the entire archive, becise it has not been published yet.

Reviewer: ```suggestion
Meaning we can delete the entire archive, because it has not been published yet.
```

> Once we publish thuogh, we have a DOI, which is persistent and we can't take it back.

Reviewer: ```suggestion
Once we publish though, we have a DOI, which is persistent and we can't take it back.
```

Something I am painfullly aware of in that I have a post that was published twice under two different DOIs as I was working on the pipeline.

> I feel silly that happened, but now that it's done it's done. I can't take it back, it's there for perpetuity... or untill interwebs are no longer a thing.

Reviewer: ```suggestion
I feel silly that happened, but now that it's done it's done. I can't take it back, it's there for perpetuity... or until interwebs are no longer a thing.
```

    post_content <- c(
      post_content[1],
      sprintf("doi: %s", pub_deposition$metadata$doi),
      post_content[2:length(post_content)]
    )
  
    writeLines(post_content, post)
  }

  return(pdf_file)
}

# Zenodo API settings
zenodo_api_endpoint <- "https://zenodo.org/api/deposit/depositions"
zenodo_api_token <- Sys.getenv("ZENODO_API_TOKEN")

# Read Hugo content files
posts <- list.files(
  "content/blog", 
  pattern = "^index\\.md$", 
  recursive = TRUE,
  full.names = TRUE
) 

# Only process files without doi and that are published
posts <- posts[sapply(posts, needs_doi)]

# Run thorugh all posts that need a doi.
sapply(posts,
  publish_to_zenodo, 
  upload = FALSE
)

```

That is the entire process.
Now, I just need to have it working on GitHub Actions, which I won't be able to truly test before this post goes live.
Wish me luck!!

I'm happy now that I can archive all my posts in perpetuity like this, making it easier for people in the future to find them again even if my website goes down. 
I don't know how useful they will be for long, but they are there at least!
Also, they are not possible to find through other means than just search engines like google. 

> Being archived on a platform that is for science output of verious types means more people may be able to find them.

Reviewer: ```suggestion
Being archived on a platform that is for science output of various types means more people may be able to find them.
```
